I2c ref impl

maciejbocianski opened the pull request:
### Description

This PR is a continuation of the work done in #8682.
It contains all changes from #8682 plus:
- hal-i2c slave mode fix for k64f
- hal-i2c async mode implementation for k64f
- I2CSlave driver adjustment to new HAL I2C API
- async implementation in I2C driver

This PR contains reference implementations of the I2C HAL RFC changes for the FRDM-K64F and NUCLEO_F070RB boards.

- [x] TasksK64F Sync
- [x] STM32 Sync
- [x] STM32 Async
- [x] Driver API Sync
- [x] Driver API Async

### Pull request type

<!--
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [X] Breaking change

### Reviewers

@ARMmbed/mbed-os-hal 
<!--
    Optional
    Request additional reviewers with @username
-->

### Release Notes

<!--
    Optional
    In case of breaking changes, functionality changes or refactors, please add release notes here. 
    For more information, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html#pull-request-types).
-->

Commit:
docs(hal-i2c): Add initial design document for the I2C HAL API overhaul

Commit:
docs(hal-i2c): Add modifications for timeout and multimaster support

Commit:
docs(hal-i2c): Update the RFC to reflect removing the DMAUsage parameter

Commit:
docs(hal-i2c): Add missing stop and last parameters to i2c_read/write

Commit:
docs(hal-i2c): Move address parameter into a seperate i2c_slave function

Commit:
docs(hal-i2c): Add additional defined function behaviours

Commit:
feat(hal-i2c): Add asynchronous HAL I2C implementation for STM32

Commit:
docs(hal-i2c): Make additional amendments to I2C API specification

- Add return value to i2c_frequency indicating configured frequency.
- Change return type of i2c_write/i2c_read from int to int32_t
- Add type suffix to i2c_slave_status enum.
- Rename i2c_slave_receive to i2c_slave_status.

Commit:
docs(hal-i2c): Add additional notes to synchronous read/write

Commit:
docs(hal-i2c): Add clock stretching field to target capabilities

Commit:
feat(hal-i2c): Update the HAL I2C header file to new API

Commit:
feat(hal-i2c): Disable I2C on all targets in preparation

Commit:
feat(hal-i2c): Add synchronous HAL I2C implementation for K64F

Commit:
feat(hal-i2c): Update I2C driver API to new specification

Commit:
feat(hal-i2c): Add synchronous HAL I2C implementation for STM32

Commit:
docs(hal-i2c): Add context pointer to async transfer function

Commit:
docs(hal-i2c): Fix return type of i2c_frequency function

Commit:
feat(hal-i2c): Remove unused DMAUsage member from I2C class

Commit:
fix(hal-i2c): Fix issue with master byte read/write failing on K64F

Remove byte_read/byte_write functions and use block read and writes for
single byte accesses.

Commit:
fix(hal-i2c): Add missing i2c_free function to API

Commit:
docs(hal-i2c): Move design document to the updated folder

Commit:
fix(hal-i2c): Fix incorrect minimum frequency in STM I2C implementation

Commit:
fix(hal-i2c): Fix STM implementation calling event handler too early

An async transfer involving a write and a read is issue in two stages,
the write is invoked, then in the write complete callback the read is
invoked followed by the read complete callback.

Currently the read and write complete callbacks invoke the HAL API event
handler, but this will be incorrectly called halfway through the
transfer after the write is completed.

To fix, only call the HAL event handler in the write complete callback
if there is no pending read.

Commit:
fix(hal-i2c): Return the correct number of transferred bytes in STM32

The STM32 attribute XferCount does not contain the number of transferred
bytes, it contains the number of bytes remaining to be sent. Calculate
the number of bytes that have been sent by subtracting this number from
the length of the respective buffer.

To fix, add an additional attribute to the asynchronous event object.
For transfers that include sending and receiving data only counting the
number of bytes sent isn't sufficient as unlike SPI the size of the send
and receive buffer may be different.

Commit:
refactor(hal-i2c): Rename HAL slave status enum to match Driver API

Commit:
fix(hal-i2c): Fix incorrect byte counts during aborted R/W transfers

The async event callback returns the number of bytes sent and received
during a transfer. For failed transfers that involve a write and read we
don't easily know which part of the transfer fails, only the number of
bytes remaining in the current operation.

To fix, set a flag when the write completes successfully, a failure
when this flag is set means the read stopped early and we can calculate
the number of bytes sent for each operation.

Commit:
refactor(hal-i2c): Change STOP function to return success or failure

Commit:
fix(hal-i2c): Add missing semicolon to STM32 I2C attribute 'tx_count'

Commit:
fix(hal-i2c): STM: Return the num of bytes sent for incomplete transfers

Commit:
refactor(hal-i2c): Remove return status from I2C start/stop functions

The return status doesn't make sense in this context, and most platforms
can't detect failure at this point. Platforms that do support detecting
failure here can handle it during read/write anyway.

Commit:
docs(hal-i2c): Update design document to match recent API alterations

Commit:
fix(hal-i2c): STM: Fix frequency function checking the wrong variable

The frequency function checks the frequency requested by the caller,
rather than the actual supported frequency selected by the function.

Commit:
style(hal-i2c): STM: Remove empty comment block from code

Commit:
fix(hal-i2c): STM: Add missing return statement when re-initing as slave

- Stop function should exit early if the device is reinitialised in
  slave mode.

- Remove slave mode lookup when reinitialising, this block is only hit
  if the peripheral is in slave mode, we don't need to check again.

Commit:
refactor(hal-i2c): Add an API function to configure clock stretching

- Remove clock stretching behavior from `i2c_timeout` function and
  define a new function to enable/disable clock stretching behavior.

- Add implementation for STM32 boards and track configuration in the
  platform specific I2C struct.

Commit:
refactor(hal-i2c): Remove i2c_timeout function from the API

Commit:
style(hal-i2c): Fix astyle code formatting failures

Commit:
fix(hal-i2c): K64F: Ignore calls to set slave address in master mode

Commit:
fix(hal-i2c): Disable `I2CSLAVE` feature flag on K64F devices

Commit:
fix(hal-i2c): Return the num bytes transferred in STM slave read/write

Commit:
fix(hal-i2c): Disable slave mode and 10-bit from I2C capabilities on K64

Commit:
fix(hal-i2c): Disable 10-bit addressing I2C capabilities on STM32

Commit:
fix(hal-i2c): Fix compiler warning from unused variable on STM32

Commit:
fix(hal-i2c): Fix return value for transfers that timeout in STM

maciejbocianski committed:
rebase fixes

maciejbocianski committed:
fix(hal-i2c): fix slave mode for k64f

maciejbocianski committed:
imp(hal-i2c): async implementation for k64f

maciejbocianski committed:
impl(driver-i2c): adjust I2CSlave driver to new HAL I2C API

maciejbocianski committed:
impl(driver-i2c): async implementation for I2C driver

maciejbocianski committed:
fix(driver-i2c): astyle fixes

ithinuel commented on targets/targets.json:
```diff
@@ -5306,6 +5186,42 @@
             "target.network-default-interface-type": "ETHERNET"
         }
     },
+    "ARM_BEETLE_SOC": {
```
Is this intended to be here ?  It is not mentioned on the PR description.

maciejbocianski replied:
fixed

ithinuel commented on drivers/I2C.cpp:
```diff
@@ -41,16 +42,23 @@ I2C::I2C(PinName sda, PinName scl) :
     recover(sda, scl);
-    i2c_init(&_i2c, _sda, _scl);
+    i2c_init(&_i2c, sda, scl, false);
+    i2c_frequency(&_i2c, _hz);
```
`I2C::frequency` directly from here to avoid code duplication ?

maciejbocianski replied:
fixed

ithinuel commented on drivers/I2C.cpp:
```diff
@@ -32,7 +32,8 @@ SingletonPtr<PlatformMutex> I2C::_mutex;
 
 I2C::I2C(PinName sda, PinName scl) :
```
This implementation does not seem to deal with multiple i2c peripheral used in parallel very well.

- Only 1 I2C can be used at a time.
- There is no mechanism that I can see that prevent an i²c peripheral to be instantiated twice.
- The `I2C` class seem to be oriented to have `1 instance per bus` instead of `1 instance per slave` like `SPI` class does.
  This is probably out of the scope of this PR as it'd create breaking change on the C++ API level but it is still worth nothing that this is inconsistent and should be fixed.

maciejbocianski replied:
I agree the improvement of `I2C` and `I2CSlave` drivers should be in separate PR. This PR should include only adaptation to HAL API changes

donatieng replied:
Yes, adding extra functionality to the driver is out of scope

ithinuel commented on docs/design-documents/hal/0001-i2c-overhaul.md:
```diff
@@ -0,0 +1,494 @@
+ *  @param tx_length The number of bytes to transmit
+ *  @param rx        The receive buffer
+ *  @param rx_length The number of bytes to receive
+ *  @param address   The address to be set - 7bit or 9bit
```
It should read `10bit` rather than `9bit`

ithinuel commented on docs/design-documents/hal/0001-i2c-overhaul.md:
```diff
@@ -0,0 +1,494 @@
+
+- **Change** the `address` parameter in `i2c_write` and `i2c_read` functions from `int` to `uint16_t`.
+
+  The address parameter is up to a 9-bit value, specifying a type with correct sign and size is sensible.
```
Address parameter is up to 10-bit not 9.

maciejbocianski replied:
fixed

ithinuel commented on docs/design-documents/hal/0001-i2c-overhaul.md:
```diff
@@ -0,0 +1,494 @@
+
+  The length parameter cannot be signed.
+
+- **Change** the `data` parameter in `i2c_write` and `i2c_read` functions from `char*` to `void*`.
```
Since the symbol size if always 8bits it would probably be sensible to use `uint8_t*` or (`int8_t*`) to eliminate any endianness related issues/unexpected behaviour.

ithinuel commented on docs/design-documents/hal/0001-i2c-overhaul.md:
```diff
@@ -0,0 +1,494 @@
+
+  Currently the `DMAUsage` argument of the `i2c_transfer_asynch` function is unimplemented, the argument is unused by all `I2C` implementations. There is no real demand for it so can be removed from the API.
+
+- **Change** the `stop` parameter from the `i2c_transfer_async` from an `uint32_t` to a `bool`.
```
Why isn't this argument removed Since it is also removed from the sync API ?

maciejbocianski replied:
Actually `stop` stays for both sync and async API

ithinuel commented on hal/i2c_api.h:
```diff
@@ -44,13 +46,26 @@
 /**@}*/
 
 #if DEVICE_I2C_ASYNCH
```
I am not super fan of this choice of design as it makes it required to use a macro to access the inner `i2c_s` and that IMHO impair readability and maintainability.

Why not simply doing :
```c
[...]
typedef struct i2c_s i2c_t;
typedef void (*i2c_async_handler_f)(i2c_t *obj, i2c_async_event_t *event, void *ctx);
struct i2c_s {
    struct target_i2c_s i2c;        /**< Target specific I2C structure */
#if DEVICE_I2C_ASYNCH
    struct buffer_s tx_buff; /**< Tx buffer */
    struct buffer_s rx_buff; /**< Rx buffer */
    i2c_async_handler_f handler;
    void *ctx;
#endif
};
```

This makes the async support transparent to the rest of the code and ease the reading for a developer new to this API.

maciejbocianski replied:
It was inherited from old implementation

maciejbocianski replied:
Changed

ithinuel commented on hal/i2c_api.h:
```diff
@@ -61,91 +76,170 @@ typedef struct i2c_s i2c_t;
  *  @param data    The buffer for sending
  *  @param length  Number of bytes to write
- *  @param stop    Stop to be generated after the transfer is done
+ *  @param stop    If true, stop will be generated after the transfer is done
```
Shouldn't this be removed from the parameter list as per the design doc ?

maciejbocianski replied:
`stop` stays

ithinuel commented on hal/i2c_api.h:
```diff
@@ -61,91 +76,170 @@ typedef struct i2c_s i2c_t;
+ *  @param address 7-bit address (last bit is 1)
+ *  @param data    The buffer for receiving
+ *  @param length  Number of bytes to read
+ *  @param last    If true, indicates that the transfer contains the last byte
```
This seems to be an equivalent to `stop` so shouldn't this be removed from the parameter list as per the design doc ?

maciejbocianski replied:
fixed

ithinuel reviewed (changes requested):
- There are inconsistencies on address' maximum lenght (9/10bits)
- The API snippet of the design doc is inconsistent with the statements made before (eg. `stop` argument to `i2c_read` and `i2c_write`).
- Why Async has a single "transfer" while sync has "read" and "write" ?
- What is the other of read/write operations in Async ? 
- How the device is expected to process an abort of the transfer ?
  - Should it send a stop if one was requested ?
	- Are symbols atomic ? That is to say, should the transfer stop in the middle of a transmission when abort is called or should it gracefully stop after the current symbol ?
	- Is an abort considered an error and should it be reported as such in the event ?
	- Is the event callback invoked at all in case of abort ?
	
There are also typos here and there but they're minor issues in regards to the comment above & in the code.

I have not reviewed (yet) the implementations as I don't know (yet) the platform, my feedback on these would not be very deep. And the API should probably be fixed before we give attention to the details of the implementations.

maciejbocianski committed:
update I2C HAL API/design doc

- update return values and parameter types of for few functions
- update i2c_timeout function specification
- fix minor typos

maciejbocianski committed:
STM I2C HAL implementation update

- add timeout for blocking transmission
- remove old code
- sync with API changes

maciejbocianski committed:
k64f I2C HAL implementation update

- add timeout for blocking transmission
- sync with API changes
- revert unnecessary k64f SDK driver changes

maciejbocianski committed:
adjust I2C and I2CSlave driver to HAL I2C API chnges

- add timeout for blocking transmission
- sync with HAL I2C API changes

jamesbeyond reviewed (approved):
I am happy about the changes

maciejbocianski committed:
HAL i2c structure update

maciejbocianski committed:
HAL i2c timeout functinality update

maciejbocianski committed:
HAL i2c design doc update

maciejbocianski committed:
HAL i2c: add multi-master support and doc update

- update I2C API and doc to include multimaster arbitration
- extend error signaling by arbitration lost error code
- after review update doc
- update k64f and STM reference implementation according API changes

donatieng commented on docs/design-documents/hal/0001-i2c-overhaul.md:
```diff
@@ -0,0 +1,527 @@
+
+- **Add** two more error status.
+
+  `I2C_ERROR_TIMEOUT` returned by `i2c_write` and `i2c_read` when transminssion timeout occurs.
```
```suggestion
  `I2C_ERROR_TIMEOUT` returned by `i2c_write` and `i2c_read` when transmission timeout occurs.
```

maciejbocianski replied:
fixed

donatieng commented on docs/design-documents/hal/0001-i2c-overhaul.md:
```diff
@@ -0,0 +1,527 @@
+
+- **Change** the `stop` parameter from the `i2c_transfer_async` from an `uint32_t` to a `bool`.
+
+  The stop parameter indicates whether or not the function should send a `STOP` command after the transfer has complete, there is no reason at all for this to be a `uint32_t`.
```
```suggestion
  The stop parameter indicates whether or not the function should send a `STOP` command after the transfer has completed, there is no reason at all for this to be a `uint32_t`.
```

maciejbocianski replied:
fixed

donatieng commented on docs/design-documents/hal/0001-i2c-overhaul.md:
```diff
@@ -0,0 +1,527 @@
+
+- **Change** the return type of `i2c_transfer_asynch` to indicate whether or not a transfer has been scheduled or not.
+
+  The function now returns a `bool` indicate if a transfer was scheduled or not, which can occur if the peripheral is already busy.
```
```suggestion
  The function now returns a `bool` indicating if a transfer was scheduled or not, which can occur if the peripheral is already busy.
```

donatieng commented on docs/design-documents/hal/0001-i2c-overhaul.md:
```diff
@@ -0,0 +1,527 @@
+
+- **Remove** the `i2c_active` function from the API.
+
+  The the async callback is now always invoked on async operation  termination (unless cancelled), this status can be tracked from driver layer without any HAL request.
```
```suggestion
  The async callback is now always invoked on async operation termination (unless cancelled), this status can be tracked from driver layer without any HAL request.
```

donatieng commented on docs/design-documents/hal/0001-i2c-overhaul.md:
```diff
@@ -0,0 +1,527 @@
+    /**< Maximum frequency supported must be set by target device */
+    uint32_t    maximum_frequency;
+    /**< If true, the device can handle I2C slave mode. */
+    bool      supports_slave_mode;
```
```suggestion
    bool      supports_slave_mode : 1;
```

This field and the ones below can be packed to save a bit of memory.

donatieng commented on docs/design-documents/hal/0001-i2c-overhaul.md:
```diff
@@ -0,0 +1,527 @@
+    I2C_ERROR_ARBITRATION_LOST = -4
+};
+
+typedef struct {
```
Is it worth packing this struct?

maciejbocianski replied:
I'm not sure. Will save only `3B` per instance. An further savings will depend on how this struct will be utilized in reworked I2C drivers layer.

donatieng commented on docs/design-documents/hal/0001-i2c-overhaul.md:
```diff
@@ -0,0 +1,527 @@
+ * @param enabled If 'true' enable clock stretching on the given I2C peripheral,
+ *                otherwise disable it.
+ */
+void i2c_set_clock_stretching(i2c_t *obj, const bool enabled);
```
No reason to make `bool` `const`

donatieng commented on docs/design-documents/hal/0001-i2c-overhaul.md:
```diff
@@ -0,0 +1,527 @@
+ */
+void i2c_slave_address(i2c_t *obj, uint16_t address);
+
+typedef void (*i2c_async_handler_f)(i2c_t *obj, i2c_async_event_t *event, void *ctx);
```
Where is `i2c_async_event_t` defined?

maciejbocianski replied:
fixed

donatieng commented on docs/design-documents/hal/0001-i2c-overhaul.md:
```diff
@@ -0,0 +1,527 @@
+typedef void (*i2c_async_handler_f)(i2c_t *obj, i2c_async_event_t *event, void *ctx);
+
+/** Start I2C asynchronous transfer
+ *
```
This needs more doc on how this is used

maciejbocianski replied:
fixed

donatieng commented on docs/design-documents/hal/0001-i2c-overhaul.md:
```diff
@@ -0,0 +1,527 @@
+  - Handles transfer collisions and loss of arbitration if the platform supports multimaster in hardware.
+  - `i2c_async_event_t` must be filled with the number of symbols sent to the bus during transfer.
+- `i2c_abort_async`:
+  - Aborts any on-going async transfers.
```
This should be optional - "best effort".

maciejbocianski replied:
changed

donatieng commented on hal/i2c_api.h:
```diff
@@ -39,113 +41,227 @@
+ *                otherwise disable it.
  */
-int  i2c_stop(i2c_t *obj);
+void i2c_set_clock_stretching(i2c_t *obj, const bool enabled);
```
```suggestion
void i2c_set_clock_stretching(i2c_t *obj, bool enabled);
```

donatieng commented on hal/i2c_api.h:
```diff
@@ -192,43 +308,28 @@ const PinMap *i2c_slave_scl_pinmap(void);
- *  @return non-zero if a value is available
- */
-void i2c_slave_mode(i2c_t *obj, int enable_slave);
+typedef enum {
```
Needs more doc - what's the default status? Should `NoData` be `Idle`?

maciejbocianski replied:
Yes, `Idle` will be more suitable here

maciejbocianski replied:
fixed

donatieng commented on hal/i2c_api.h:
```diff
@@ -39,113 +41,227 @@
+ *  @param stop    If true, stop will be generated after the transfer is done
+ *
+ *  @note If the current platform supports multimaster operation the peripheral
+ *        will perform arbitration automatically when detects collision and
```
```suggestion
 *        will perform arbitration automatically when detecting collisions and
```

donatieng commented on hal/i2c_api.h:
```diff
@@ -39,113 +41,227 @@
+ *        complete the transfer or return I2C_ERROR_ARBITRATION_LOST
+ *        when loses arbitration.
+ *
+ *        Additional time for arbitration or clock stretching should by count
```
How much time?

maciejbocianski replied:
Actually we don't know how long it would take. It's up to user to decide how much time he can wait before failing

donatieng commented on targets/TARGET_Freescale/TARGET_MCUXpresso_MCUS/api/i2c_api.c:
```diff
@@ -24,313 +25,646 @@
-    base->C1 &= ~(I2C_C1_TX_MASK | I2C_C1_TXAK_MASK);
+    /* Add this to avoid build warning. */
+    dummy++;
+    //printf("_I2C_SlaveWriteBlocking timeout1: %u\n", timeout);
```
Should be removed

donatieng commented on targets/TARGET_Freescale/TARGET_MCUXpresso_MCUS/api/i2c_api.c:
```diff
@@ -24,313 +25,646 @@
-        base->S = kI2C_ArbitrationLostFlag;
-        ret_value = 2;
-    }
+    //printf("_I2C_SlaveWriteBlocking timeout2: %u\n", timeout);
```
Same

donatieng reviewed (changes requested):
Thanks @maciejbocianski - looks good, some small changes/clarifications required in some places

maciejbocianski committed:
HAL i2c: fix doc and API typos

maciejbocianski committed:
HAL i2c: docs and API updates

After review, API and documentation update:
  - change slave status enum NoData -> Idle
  - minor documentation update

maciejbocianski committed:
Driver i2c: update timeout documentation

ithinuel commented on drivers/I2C.cpp:
```diff
@@ -41,16 +42,23 @@ I2C::I2C(PinName sda, PinName scl) :
 {
     lock();
-    _hz = hz;
+    _hz = (uint32_t)hz;
```
*nitpick:* Should this be guarded to prevent setting of negative values ?

maciejbocianski replied:
Fixed
`MBED_ASSERT(_hz > 0);` added for both master and slave driver

ithinuel reviewed (approved):
LGTM 👍  Thanks for the huge work 😃

maciejbocianski committed:
Driver i2c: improve frequency setting function

maciejbocianski committed:
LPC408X: temporarily disable ETHERNET API to allow I2C-less builds

maciejbocianski committed:
UBLOX_C030: temporarily disable I2C API and min_battery_voltage to allow I2C-less builds

maciejbocianski committed:
EFM32: remove DEVICE_I2C guard from I2CName

I2CName should stay when I2C is disabled, becouse it's required by PeripheralPins.c

maciejbocianski committed:
STM32WB: fix i2c_s struct

maciejbocianski committed:
STM32H7: fix i2c_s struct

maciejbocianski committed:
GR_LYCHEE: temporarily disable TRNG since it uses I2C

maciejbocianski committed:
Disable examples unsupported on i2c feature branch.
